wal/wal.go

```diff
@@ -615,17 +619,21 @@ func (w *WAL) log(rec []byte, final bool) error {
 		copy(buf[recordHeaderSize:], part)
 		p.alloc += len(part) + recordHeaderSize
 
-		// By definition when a record is split it means its size is bigger than
-		// the page boundary so the current page would be full and needs to be flushed.
-		// On contrary if we wrote a full record, we can fit more records of the batch
-		// into the page before flushing it.
-		if final || typ != recFull || w.page.full() {
-			if err := w.flushPage(false); err != nil {
+		if w.page.full() {
```

Comment:
hm , this doesn't look right, There can be a case where the page is not full, but the next record won't fit so needs to e flushed.

Reply:
`typ == recFirst || typ == recMiddle` means that the page is full, so a single check should be enough, isn't it?

Looking at the code I think it is ensured that at least the header can be put for the next record. I know putting only header sounds inefficient, but that was the case before too I believe on some occassions if `typ` was `recLast`.

Reply:
@krasi-georgiev @codesome @csmarchbanks 

If typ == recFirst || type == recMiddle, then w.page.full() is engouh actually.

But if typ is recFull or recLast and the remaining space after writing this record is < recordHeaderSize. Then for next record, `l = min(len(rec), (pageSize-p.alloc)-recordHeaderSize)` will < 0 and `part=rec[:l]` will result panic.

So atfer write a full (or part of ) the record and w.page.alloc <= recordHeaderSize, then this page should be flushed.

Reply:
How does the panic happen?
* If there is not enough space to write a header, then `w.page.full()` will be true and the page will flush. 
* If there is exactly enough space to write a header then `part=rec[:0]`, which would wastefully write an empty part, but not panic.

If `w.page.full()` is not enough for `recFull`, wouldn't there be panics in the master code?

Reply:
@csmarchbanks My mistake, I understand page.full() literally :joy: 

As @csmarchbanks said, the code in this PR has same effect as the master code.

If the code in this PR is wrong, then the master code is wrong as well.

Reply:
Nope, it does not panic. The explanation is precisely what @csmarchbanks said.

To make it less wasteful should we also check for some X bytes to be available in addition to headers before writing the record part? Not sure what that X has to be.

Reply:
Making it less wasteful would be nice, but I would say it should probably be a separate PR.